### PR TITLE
JVM_IR: support @JvmDefaults in compatibility mode.

### PR DESCRIPTION
**Pull request by ting-yuan**
It's easier to start with a small PR. Some TODOs to be done in later PRs.
1. -Xjvm-default=enabled will be implemented later.
2. There are also a dozen of related tests failing and need to be fixed with other phases, such as accessor.

There are some changes in BridgeLowering.kt because otherwise it breaks some existing tests.

### CODE REVIEW

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt [hunk 1]
```diff
@@ -67,6 +80,33 @@ private class InterfaceLowering(val context: JvmBackendContext) : IrElementTrans
         Visibilities.isPrivate(function.visibility) && function.name != clinitName ||
                 function.origin == IrDeclarationOrigin.FUNCTION_FOR_DEFAULT_PARAMETER ||
                 function.origin == JvmLoweredDeclarationOrigin.SYNTHETIC_METHOD_FOR_PROPERTY_ANNOTATIONS
+
+    private fun IrFunction.isJvmDefault(): Boolean =
+        (hasAnnotation(JVM_DEFAULT_FQ_NAME) ||
+                (this is IrSimpleFunction && correspondingPropertySymbol?.owner?.hasAnnotation(JVM_DEFAULT_FQ_NAME) == true)) &&
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt]: 'propertyIfAccessor' could be used here

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt [hunk 2]
```diff
@@ -67,6 +80,33 @@ private class InterfaceLowering(val context: JvmBackendContext) : IrElementTrans
...
+
+    private fun callDefaultImpls(defaultImpls: IrFunction, interfaceMethod: IrFunction): IrCall {
+        val startOffset = interfaceMethod.startOffset
+        val endOffset = interfaceMethod.endOffset
+
+        return IrCallImpl(interfaceMethod.startOffset, interfaceMethod.endOffset, interfaceMethod.returnType, defaultImpls.symbol).apply {
+            passTypeArgumentsFrom(interfaceMethod)
+
+            var offset = 0
+            interfaceMethod.dispatchReceiverParameter?.let {
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt]:
```        
       val receiverAndArgs = interfaceMethod.receiverAndArgs()
        receiverAndArgs.forEachIndexed { i, irExpression ->
        newExpression.putValueArgument(i, irExpression)
```
        }

**Comment by max-kammerer** [comment]
@ting-yuan Please move bridge related changes to separate pr, there are some additions:
1. Please note that with these changes bridge is generated twicely in interface and in inherited class: see bridgeInInterface2.kt test
2.  '!it.owner.comesFromJava()' check in 'createBridges' fun looks like wrong, there should be 'it.owner.comesFromJava() || it.hasJvmDefault()'
3. Looks like 'irFunction.overriddenSymbols.all ...' should be substituted with ```irFunction.findConcreteSuperDeclaration()```

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt (hunk 2 above)
Comment (ting-yuan) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt]: `.receiverAndArgs()` seems only available for `IrFunctionAccessExpression` (call site) but not `IrFunction` (definition). Looks like `IrGetValueImpl()` has to be called explicitly?

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt (hunk 1 above)
Comment (ting-yuan) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt]: I searched `propertyIfAccessor` and all of them seem to be extensions of some descriptors?

**Comment by ting-yuan** [comment]
@max-kammerer Done splitting out bridge related changes. Will upload it in a separate PR.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt (hunk 1 above)
Comment (max-kammerer) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt]: Oh...seems we should add one for  IrFunction

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt (hunk 2 above)
Comment (max-kammerer) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt]: Yes, sorry for mistake

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt (hunk 1 above)
Comment (ting-yuan) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt]: @max-kammerer Done! Please take a look again.

**Comment by ting-yuan** [comment]
@max-kammerer I also moved `IrFunction.isJvmDefault()` to `jvm/ir/IrUtils.kt` so that it can be used by other phases: the check `body == null` doesn't work when referencing other compilation units.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/ir/IrUtils.kt [hunk 3]
```diff
@@ -37,3 +41,17 @@ val IrType.erasedUpperBound: IrClass
...
         else -> throw IllegalStateException()
     }
+
+val IrFunction.propertyIfAccessor: IrDeclaration
+    get() = (this as? IrSimpleFunction)?.correspondingPropertySymbol?.owner ?: this
+
+fun IrFunction.annotatedWithJvmDefault(): Boolean = propertyIfAccessor.hasAnnotation(JVM_DEFAULT_FQ_NAME)
+
+fun IrFunction.isJvmDefault(context: JvmBackendContext): Boolean = annotatedWithJvmDefault() &&
+        context.state.target.bytecodeVersion >= JvmTarget.JVM_1_8.bytecodeVersion &&
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/ir/IrUtils.kt]: Version check is performed in frontend and here it's redundant: IrFunction could be from another module compiled to 1.8 and this one is compiling to 1.6

Comment (max-kammerer) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/ir/IrUtils.kt]: jvmDefaultMode.isEnabled check is also performed in frontend

Comment (ting-yuan) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/ir/IrUtils.kt]: Thanks for pointing that out. I've updated it accordingly.

**Comment by max-kammerer** [comment]
Thanks!

**Pull request merged**